assets/scripts/segments/Segment.jsx

```diff
@@ -69,35 +60,28 @@ class Segment extends React.Component {
   componentDidUpdate (prevProps, prevState) {
     if (this.props.forPalette) return
 
-    this.drawSegment(this.props.variantString, false)
-
     if (prevProps.suppressMouseEnter && !this.props.suppressMouseEnter &&
         infoBubble.considerSegmentEl === this.streetSegment) {
       infoBubble.considerShowing(false, this.streetSegment, INFO_BUBBLE_TYPE_SEGMENT)
     }
 
-    if (prevProps.variantString !== this.props.variantString) {
+    if (prevProps.variantString && prevProps.variantString !== this.props.variantString) {
       this.switchSegments(prevProps.variantString)
     }
 
-    if (this.state.switchSegments && prevState.switchSegments !== this.state.switchSegments) {
-      this.drawSegment(this.state.oldSegmentVariant, true)
-      this.updateOldCanvasLeftPos(this.state.oldSegmentVariant)
-      this.props.updatePerspective(this.oldSegmentCanvas)
-      this.props.updatePerspective(this.segmentCanvas)
+    if (!prevState.switchSegments && this.state.switchSegments) {
+      this.props.updatePerspective(this.oldSegmentCanvas.firstChildElement)
+      this.props.updatePerspective(this.newSegmentCanvas.firstChildElement)
     }
 
     this.props.updateSegmentData(this.streetSegment, this.props.dataNo, this.props.segmentPos)
   }
 
-  drawSegment = (variantString, isOldSegment) => {
-    const multiplier = this.props.forPalette ? (WIDTH_PALETTE_MULTIPLIER / TILE_SIZE) : 1
-    const segmentWidth = this.props.width // may need to double check this. setSegmentContents() was called with other widths
-    const offsetTop = this.props.forPalette ? SEGMENT_Y_PALETTE : SEGMENT_Y_NORMAL
-    const canvas = (isOldSegment) ? this.oldSegmentCanvas : this.segmentCanvas
-    const ctx = canvas.getContext('2d')
-    ctx.clearRect(0, 0, canvas.width, canvas.height)
-    drawSegmentContents(ctx, this.props.type, variantString, segmentWidth, 0, offsetTop, this.props.randSeed, multiplier, this.props.forPalette)
+  switchSegments = (oldVariant) => {
```

Comment:
One thing we can do in this component now is introduce a [`componentDidCatch()`](https://reactjs.org/docs/error-boundaries.html) error boundary here. This way if anything goes wrong during the render in `<SegmentCanvas />` we can handle an error inside this component itself. Currently if a render goes wrong in a segment, it crashes the entire app.

Reference #1039. (Ideally, we render the "unknown" placeholder data -- https://github.com/streetmix/streetmix/blob/1511e00aefe2026346573d3022f633bf239936a8/assets/scripts/segments/info.js#L19-L42)

Reply:
Sorry, how do you want to render the placeholder data when there is an error? Like what exactly do we display? Are we still calling `drawSegmentContents`?

Reply:
Yeah - the gist of this is that there is a segment defined as "unknown" and so that is what gets rendered instead. On further consideration I'm not positive that it's the best solution now because this refactoring basically allows us to draw a placeholder without reusing the SegmentCanvas component which might be better if that's the component that's failing. I think we can keep things simple, put in the componentDidCatch () now and figure out what to render instead in a followup PR.